// @twilio-paste/list/dist/index.debug.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  __markAsModule(target);
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  if (module2 && module2.__esModule)
    return module2;
  return __exportStar(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", {value: module2, enumerable: true}), module2);
};

// src/index.tsx
__export(exports, {
  List: () => List,
  ListItem: () => ListItem,
  OrderedList: () => OrderedList,
  UnorderedList: () => UnorderedList
});

// src/List.tsx
var React = __toModule(require("react"));
var PropTypes = __toModule(require("prop-types"));
var import_text = __toModule(require("@twilio-paste/text"));
var import_style_props = __toModule(require("@twilio-paste/style-props"));
var List = React.forwardRef((_a, ref) => {
  var {as, children, element = "LIST"} = _a, props = __rest(_a, ["as", "children", "element"]);
  return /* @__PURE__ */ React.createElement(import_text.Text, __assign(__assign({}, props), {
    as,
    color: "colorText",
    element,
    fontSize: "fontSize30",
    fontWeight: "fontWeightNormal",
    lineHeight: "lineHeight40",
    marginLeft: "space70",
    ref
  }), children);
});
List.displayName = "List";
List.propTypes = {
  as: PropTypes.oneOf(["ol", "ul"]).isRequired,
  element: PropTypes.string,
  marginTop: import_style_props.isMarginTokenProp,
  marginBottom: import_style_props.isMarginTokenProp
};

// src/ListItem.tsx
var React2 = __toModule(require("react"));
var PropTypes2 = __toModule(require("prop-types"));
var import_text2 = __toModule(require("@twilio-paste/text"));
var ListItem = React2.forwardRef((_a, ref) => {
  var {children, element = "LIST_ITEM"} = _a, props = __rest(_a, ["children", "element"]);
  return /* @__PURE__ */ React2.createElement(import_text2.Text, __assign(__assign({}, import_text2.safelySpreadTextProps(props)), {
    as: "li",
    color: "colorText",
    element,
    fontSize: "fontSize30",
    fontWeight: "fontWeightNormal",
    lineHeight: "lineHeight40",
    marginBottom: "space30",
    marginTop: "space30",
    ref
  }), children);
});
ListItem.displayName = "ListItem";
ListItem.propTypes = {
  element: PropTypes2.string
};

// src/OrderedList.tsx
var React3 = __toModule(require("react"));
var PropTypes3 = __toModule(require("prop-types"));
var import_text3 = __toModule(require("@twilio-paste/text"));
var import_style_props2 = __toModule(require("@twilio-paste/style-props"));
var OrderedList = React3.forwardRef((_a, ref) => {
  var {children, element = "ORDERED_LIST", marginTop, marginBottom = "space70"} = _a, props = __rest(_a, ["children", "element", "marginTop", "marginBottom"]);
  return /* @__PURE__ */ React3.createElement(List, __assign(__assign({}, import_text3.safelySpreadTextProps(props)), {
    as: "ol",
    element,
    marginTop,
    marginBottom,
    ref
  }), children);
});
OrderedList.displayName = "OrderedList";
OrderedList.propTypes = {
  element: PropTypes3.string,
  marginTop: import_style_props2.isMarginTokenProp,
  marginBottom: import_style_props2.isMarginTokenProp
};

// src/UnorderedList.tsx
var React4 = __toModule(require("react"));
var PropTypes4 = __toModule(require("prop-types"));
var import_text4 = __toModule(require("@twilio-paste/text"));
var import_style_props3 = __toModule(require("@twilio-paste/style-props"));
var UnorderedList = React4.forwardRef((_a, ref) => {
  var {children, element = "UNORDERED_LIST", marginTop, marginBottom = "space70"} = _a, props = __rest(_a, ["children", "element", "marginTop", "marginBottom"]);
  return /* @__PURE__ */ React4.createElement(List, __assign(__assign({}, import_text4.safelySpreadTextProps(props)), {
    as: "ul",
    element,
    marginTop,
    marginBottom,
    ref
  }), children);
});
UnorderedList.displayName = "UnorderedList";
UnorderedList.propTypes = {
  element: PropTypes4.string,
  marginTop: import_style_props3.isMarginTokenProp,
  marginBottom: import_style_props3.isMarginTokenProp
};
