// @twilio-paste/card/dist/index.debug.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  __markAsModule(target);
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  if (module2 && module2.__esModule)
    return module2;
  return __exportStar(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", {value: module2, enumerable: true}), module2);
};

// src/index.tsx
__export(exports, {
  Card: () => Card
});
var React = __toModule(require("react"));
var PropTypes = __toModule(require("prop-types"));
var import_box = __toModule(require("@twilio-paste/box"));
var import_style_props = __toModule(require("@twilio-paste/style-props"));
var Card = React.forwardRef((_a, ref) => {
  var {children, element = "CARD", padding = "space60", paddingBottom, paddingLeft, paddingRight, paddingTop} = _a, props = __rest(_a, ["children", "element", "padding", "paddingBottom", "paddingLeft", "paddingRight", "paddingTop"]);
  return /* @__PURE__ */ React.createElement(import_box.Box, __assign(__assign({}, import_box.safelySpreadBoxProps(props)), {
    ref,
    as: "article",
    borderWidth: "borderWidth20",
    borderColor: "colorBorderWeaker",
    borderStyle: "solid",
    borderRadius: "borderRadius20",
    element,
    padding,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
    backgroundColor: "colorBackgroundBody"
  }), children);
});
Card.propTypes = {
  element: PropTypes.string,
  padding: import_style_props.isPaddingTokenProp,
  paddingTop: import_style_props.isPaddingTokenProp,
  paddingRight: import_style_props.isPaddingTokenProp,
  paddingBottom: import_style_props.isPaddingTokenProp,
  paddingLeft: import_style_props.isPaddingTokenProp
};
Card.displayName = "Card";
