// @twilio-paste/anchor/dist/index.debug.es.js
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};

// src/index.tsx
import {
  createElement as createElement3,
  forwardRef as forwardRef3
} from "react";
import {Box as Box3, safelySpreadBoxProps} from "@twilio-paste/box";
import {LinkExternalIcon} from "@twilio-paste/icons/esm/LinkExternalIcon";

// src/DefaultAnchor.tsx
import {
  createElement,
  forwardRef
} from "react";
import {Box} from "@twilio-paste/box";
var DefaultAnchor = forwardRef((props, ref) => /* @__PURE__ */ createElement(Box, __assign(__assign({}, props), {
  as: "a",
  color: "colorTextLink",
  fontSize: "inherit",
  fontWeight: "inherit",
  lineHeight: "inherit",
  outline: "none",
  ref,
  textDecoration: "underline",
  _active: {
    color: "colorTextLinkStronger",
    textDecoration: "none"
  },
  _focus: {
    boxShadow: "shadowFocus",
    color: "colorTextLinkStronger",
    textDecoration: "none"
  },
  _hover: {
    color: "colorTextLinkStronger",
    textDecoration: "none"
  }
}), props.children));
DefaultAnchor.displayName = "DefaultAnchor";
if (false) {
  DefaultAnchor.propTypes = AnchorPropTypes;
}

// src/InverseAnchor.tsx
import {
  createElement as createElement2,
  forwardRef as forwardRef2
} from "react";
import {Box as Box2} from "@twilio-paste/box";
var InverseAnchor = forwardRef2((props, ref) => /* @__PURE__ */ createElement2(Box2, __assign(__assign({}, props), {
  as: "a",
  color: "colorTextInverse",
  fontSize: "inherit",
  fontWeight: "inherit",
  lineHeight: "inherit",
  outline: "none",
  ref,
  textDecoration: "underline",
  _active: {
    color: "colorTextInverse",
    textDecoration: "none"
  },
  _focus: {
    boxShadow: "shadowFocusInverse",
    color: "colorTextInverse",
    textDecoration: "none"
  },
  _hover: {
    color: "colorTextInverse",
    textDecoration: "none"
  }
}), props.children));
InverseAnchor.displayName = "InverseAnchor";
if (false) {
  InverseAnchor.propTypes = AnchorPropTypes;
}

// src/index.tsx
var AnchorVariants = {
  inverse: InverseAnchor,
  default: DefaultAnchor
};
var EXTERNAL_URL_REGEX = /^(https?:)\S*$/;
var EXTERNAL_TARGET_DEFAULT = "_blank";
var EXTERNAL_REL_DEFAULT = "noreferrer noopener";
var isExternalUrl = (url) => EXTERNAL_URL_REGEX.test(url);
var secureExternalLink = (href) => {
  if (!isExternalUrl(href))
    return;
  return {
    rel: EXTERNAL_REL_DEFAULT,
    target: EXTERNAL_TARGET_DEFAULT
  };
};
var Anchor = forwardRef3((_a, ref) => {
  var {
    element = "ANCHOR",
    variant = "default",
    showExternal,
    display,
    height,
    minHeight,
    maxHeight,
    width,
    minWidth,
    maxWidth,
    size,
    margin,
    marginBottom,
    marginLeft,
    marginRight,
    marginTop,
    marginX = null,
    marginY = null,
    padding,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
    paddingX = null,
    paddingY = null,
    verticalAlign
  } = _a, props = __rest(_a, [
    "element",
    "variant",
    "showExternal",
    "display",
    "height",
    "minHeight",
    "maxHeight",
    "width",
    "minWidth",
    "maxWidth",
    "size",
    "margin",
    "marginBottom",
    "marginLeft",
    "marginRight",
    "marginTop",
    "marginX",
    "marginY",
    "padding",
    "paddingBottom",
    "paddingLeft",
    "paddingRight",
    "paddingTop",
    "paddingX",
    "paddingY",
    "verticalAlign"
  ]);
  const AnchorComponent = AnchorVariants[variant];
  return /* @__PURE__ */ createElement3(AnchorComponent, __assign(__assign(__assign({
    href: props.href,
    ref,
    variant
  }, secureExternalLink(props.href)), safelySpreadBoxProps(props)), {
    element,
    display,
    height,
    minHeight,
    maxHeight,
    width,
    minWidth,
    maxWidth,
    size,
    margin,
    marginBottom,
    marginLeft,
    marginRight,
    marginTop,
    marginX,
    marginY,
    padding,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
    paddingX,
    paddingY,
    verticalAlign
  }), showExternal ? /* @__PURE__ */ createElement3(Box3, {
    as: "span"
  }, props.children, /* @__PURE__ */ createElement3(Box3, {
    as: "span",
    display: "inline-block",
    flexShrink: 0,
    verticalAlign: "middle"
  }, /* @__PURE__ */ createElement3(LinkExternalIcon, {
    decorative: false,
    title: "link takes you to an external page"
  }))) : props.children);
});
Anchor.displayName = "Anchor";
if (false) {
  Anchor.propTypes = AnchorPropTypes;
}
export {
  Anchor,
  isExternalUrl,
  secureExternalLink
};
