// @twilio-paste/stack/dist/index.debug.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  __markAsModule(target);
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  if (module2 && module2.__esModule)
    return module2;
  return __exportStar(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", {value: module2, enumerable: true}), module2);
};

// src/index.tsx
__export(exports, {
  Stack: () => Stack,
  getStackChildMargins: () => getStackChildMargins,
  getStackDisplay: () => getStackDisplay,
  getStackStyles: () => getStackStyles
});
var React = __toModule(require("react"));
var PropTypes = __toModule(require("prop-types"));
var import_uid_library = __toModule(require("@twilio-paste/uid-library"));
var import_style_props = __toModule(require("@twilio-paste/style-props"));
var import_box = __toModule(require("@twilio-paste/box"));
var getStackDisplay = (orientation) => {
  if (Array.isArray(orientation)) {
    return orientation.map((value) => {
      if (value === "horizontal") {
        return "flex";
      }
      return "block";
    });
  }
  if (orientation === "horizontal") {
    return "flex";
  }
  return "block";
};
var getStackStyles = (orientation) => {
  const styles = {
    display: getStackDisplay(orientation),
    alignItems: "center",
    flexWrap: "wrap"
  };
  return styles;
};
var getStackChildMargins = (orientation, spacing) => {
  let styles = {};
  if (Array.isArray(orientation)) {
    const marginRight = [];
    const marginBottom = [];
    orientation.forEach((value, i) => {
      marginRight[i] = value === "horizontal" ? spacing : "space0";
      marginBottom[i] = value === "horizontal" ? "space0" : spacing;
    });
    styles = {
      marginRight,
      marginBottom
    };
  }
  if (orientation === "horizontal") {
    styles = {marginRight: spacing};
  }
  if (orientation === "vertical") {
    styles = {marginBottom: spacing};
  }
  return styles;
};
var Stack = React.forwardRef((_a, ref) => {
  var {children, orientation, spacing, element = "STACK"} = _a, props = __rest(_a, ["children", "orientation", "spacing", "element"]);
  const [childrenCount, validChildren] = React.useMemo(() => {
    const filteredChildren = React.Children.toArray(children).filter((child) => React.isValidElement(child) || typeof child === "string");
    return [filteredChildren.length, filteredChildren];
  }, [children]);
  const stackStyles = React.useMemo(() => getStackStyles(orientation), [orientation]);
  const childMargins = React.useMemo(() => getStackChildMargins(orientation, spacing), [orientation, spacing]);
  const keySeed = import_uid_library.useUIDSeed();
  return /* @__PURE__ */ React.createElement(import_box.Box, __assign(__assign(__assign({
    element
  }, import_box.safelySpreadBoxProps(props)), stackStyles), {
    ref
  }), validChildren.map((child, index) => {
    return /* @__PURE__ */ React.createElement(import_box.Box, __assign(__assign({
      element: `${element}_CHILD`
    }, childrenCount !== index + 1 ? childMargins : null), {
      key: keySeed(`stack-${index}`)
    }), child);
  }));
});
Stack.displayName = "Stack";
Stack.propTypes = {
  orientation: import_style_props.ResponsiveProp(PropTypes.oneOf(["horizontal", "vertical"])).isRequired,
  spacing: import_style_props.isSpaceTokenProp,
  element: PropTypes.string
};
