// @twilio-paste/input/dist/index.debug.es.js
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};

// src/Input.tsx
import {
  createElement,
  forwardRef
} from "react";
import {
  bool,
  func,
  oneOf,
  string
} from "prop-types";
import {Box} from "@twilio-paste/box";
import {InputBox} from "@twilio-paste/input-box";

// src/utils.ts
var PROPS_TO_BLOCK = ["className", "style", "size", "height", "width"];
var safelySpreadFormControlProps = (props) => {
  return Object.keys(props).reduce((newProps, key) => {
    if (!PROPS_TO_BLOCK.includes(key)) {
      newProps[key] = props[key];
    }
    return newProps;
  }, {});
};

// src/Input.tsx
var InputElement = forwardRef((_a, ref) => {
  var {element} = _a, props = __rest(_a, ["element"]);
  return /* @__PURE__ */ createElement(Box, __assign({
    appearance: "none",
    as: "input",
    backgroundColor: "transparent",
    border: "none",
    borderRadius: "borderRadius20",
    boxShadow: "none",
    color: "inherit",
    cursor: (props.type === "date" || props.type === "time") && !props.readOnly && !props.disabled ? "text" : "auto",
    display: "block",
    element,
    fontFamily: "inherit",
    fontSize: "fontSize30",
    fontWeight: "fontWeightMedium",
    lineHeight: "lineHeight20",
    margin: "space0",
    outline: "none",
    paddingBottom: "space30",
    paddingLeft: "space40",
    paddingRight: "space40",
    paddingTop: "space30",
    resize: "none",
    width: "100%",
    variant: props.variant,
    ref,
    _placeholder: {
      color: props.variant === "inverse" ? "colorTextInverseWeak" : "colorTextWeak",
      fontStyle: "italic"
    },
    _focus_placeholder: {
      color: props.variant === "inverse" ? "colorTextInverseWeak" : "colorTextWeak"
    },
    _disabled: {
      color: props.variant === "inverse" ? "colorTextInverseWeaker" : "colorTextWeaker",
      cursor: "not-allowed",
      "-webkit-text-fill-color": props.variant === "inverse" ? "colorTextInverseWeaker" : "colorTextWeaker",
      "-webkit-opacity": "1"
    },
    __webkit_datetime_edit: {
      display: "flex"
    },
    __webkit_calendar_picker_indicator_hover: {
      cursor: props.readOnly || props.disabled ? "default" : "pointer"
    }
  }, props));
});
InputElement.displayName = "InputElement";
var Input = forwardRef((_a, ref) => {
  var {
    disabled,
    element = "INPUT",
    hasError,
    id,
    insertAfter,
    insertBefore,
    name,
    placeholder,
    readOnly,
    required,
    type,
    value,
    variant
  } = _a, props = __rest(_a, [
    "disabled",
    "element",
    "hasError",
    "id",
    "insertAfter",
    "insertBefore",
    "name",
    "placeholder",
    "readOnly",
    "required",
    "type",
    "value",
    "variant"
  ]);
  const typeProps = {type};
  if (type === "number") {
    typeProps.type = "text";
    typeProps.inputmode = "numeric";
    typeProps.pattern = "[0-9]*";
  }
  return /* @__PURE__ */ createElement(InputBox, {
    disabled,
    element,
    hasError,
    insertAfter,
    insertBefore,
    readOnly,
    type,
    variant
  }, /* @__PURE__ */ createElement(InputElement, __assign(__assign(__assign({
    "aria-invalid": hasError,
    "aria-readonly": readOnly
  }, safelySpreadFormControlProps(props)), typeProps), {
    disabled,
    element: `${element}_ELEMENT`,
    id,
    name,
    placeholder,
    readOnly,
    ref,
    required,
    value,
    variant
  })));
});
Input.displayName = "Input";
Input.propTypes = {
  disabled: bool,
  element: string,
  hasError: bool,
  id: string,
  name: string,
  onBlur: func,
  onChange: func,
  onFocus: func,
  placeholder: string,
  readOnly: bool,
  required: bool,
  type: oneOf(["text", "email", "hidden", "number", "password", "search", "tel", "date", "time"]).isRequired,
  value: string
};
export {
  Input,
  InputElement
};
