// @twilio-paste/heading/dist/index.debug.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  __markAsModule(target);
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  if (module2 && module2.__esModule)
    return module2;
  return __exportStar(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", {value: module2, enumerable: true}), module2);
};

// src/index.tsx
__export(exports, {
  Heading: () => Heading,
  HeadingPropTypes: () => HeadingPropTypes
});
var React = __toModule(require("react"));
var import_text = __toModule(require("@twilio-paste/text"));

// src/PropTypes.ts
var PropTypes = __toModule(require("prop-types"));
var HeadingPropTypes = {
  as: PropTypes.oneOf(["h1", "h2", "h3", "h4", "h5", "h6", "div", "label", "span"]).isRequired,
  element: PropTypes.string,
  marginBottom: PropTypes.oneOf(["space0"]),
  variant: PropTypes.oneOf([
    "heading10",
    "heading20",
    "heading30",
    "heading40",
    "heading50",
    "heading60"
  ]).isRequired,
  display: PropTypes.string
};

// src/index.tsx
function getHeadingProps(headingVariant, marginBottom) {
  switch (headingVariant) {
    case "heading10":
      return {
        marginBottom: marginBottom || "space70",
        fontSize: "fontSize90",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight90"
      };
    case "heading30":
      return {
        marginBottom: marginBottom || "space50",
        fontSize: "fontSize60",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight60"
      };
    case "heading40":
      return {
        marginBottom: marginBottom || "space40",
        fontSize: "fontSize40",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight40"
      };
    case "heading50":
      return {
        marginBottom: marginBottom || "space30",
        fontSize: "fontSize30",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight30"
      };
    case "heading60":
      return {
        marginBottom: marginBottom || "space30",
        fontSize: "fontSize20",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight20"
      };
    case "heading20":
    default:
      return {
        marginBottom: marginBottom || "space60",
        fontSize: "fontSize70",
        fontWeight: "fontWeightBold",
        lineHeight: "lineHeight70"
      };
  }
}
var Heading = React.forwardRef((_a, ref) => {
  var {as, children, display = "block", element = "HEADING", id, marginBottom, variant} = _a, props = __rest(_a, ["as", "children", "display", "element", "id", "marginBottom", "variant"]);
  return /* @__PURE__ */ React.createElement(import_text.Text, __assign(__assign(__assign({}, import_text.safelySpreadTextProps(props)), getHeadingProps(variant, marginBottom)), {
    as,
    display,
    element,
    id,
    color: "colorText",
    variant,
    ref
  }), children);
});
Heading.displayName = "Heading";
Heading.propTypes = HeadingPropTypes;
