// @twilio-paste/form-pill-group/dist/index.debug.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __assign = Object.assign;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __rest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  __markAsModule(target);
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  if (module2 && module2.__esModule)
    return module2;
  return __exportStar(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", {value: module2, enumerable: true}), module2);
};

// src/index.tsx
__export(exports, {
  FormPill: () => FormPill,
  FormPillGroup: () => FormPillGroup,
  useFormPillState: () => useFormPillState
});

// src/FormPill.tsx
var React2 = __toModule(require("react"));
var import_box2 = __toModule(require("@twilio-paste/box"));
var import_reakit_library = __toModule(require("@twilio-paste/reakit-library"));

// src/PillCloseIcon.tsx
var React = __toModule(require("react"));
var import_box = __toModule(require("@twilio-paste/box"));
var import_screen_reader_only = __toModule(require("@twilio-paste/screen-reader-only"));
var import_CloseIcon = __toModule(require("@twilio-paste/icons/cjs/CloseIcon"));
var PillCloseIcon = ({onClick}) => {
  return /* @__PURE__ */ React.createElement(import_box.Box, {
    onClick,
    _hover: {
      cursor: "pointer"
    }
  }, /* @__PURE__ */ React.createElement(import_screen_reader_only.ScreenReaderOnly, null, ". Press delete or backspace to remove"), /* @__PURE__ */ React.createElement(import_CloseIcon.CloseIcon, {
    decorative: true,
    size: "sizeIcon10"
  }));
};

// src/FormPill.tsx
var hoverStyles = {
  cursor: "pointer",
  _hover: {
    backgroundColor: "colorBackgroundPrimaryWeakest",
    boxShadow: "shadowBorderPrimaryStrong",
    color: "colorTextLinkStronger"
  },
  _selected_hover: {
    backgroundColor: "colorBackgroundPrimaryStronger",
    color: "colorTextWeakest"
  }
};
var FormPillStyles = React2.forwardRef((_a, ref) => {
  var {element = "FORM_PILL", selected = false, isHoverable} = _a, props = __rest(_a, ["element", "selected", "isHoverable"]);
  return /* @__PURE__ */ React2.createElement(import_box2.Box, __assign(__assign(__assign({}, import_box2.safelySpreadBoxProps(props)), {
    ref,
    element,
    as: "button",
    "aria-selected": selected,
    role: "option",
    type: "button",
    alignItems: "center",
    backgroundColor: "colorBackgroundStrong",
    borderRadius: "borderRadius10",
    borderWidth: "borderWidth0",
    color: "colorText",
    columnGap: "space20",
    cursor: "default",
    display: "flex",
    fontFamily: "fontFamilyText",
    fontSize: "fontSize20",
    fontWeight: "fontWeightSemibold",
    lineHeight: "lineHeight10",
    outline: "none",
    paddingX: "space30",
    paddingY: "space20",
    transition: "background-color 100ms ease-in, box-shadow 100ms ease-in, color 100ms ease-in, border-color 100ms ease-in",
    _focusVisible: {
      backgroundColor: "colorBackgroundPrimaryWeakest",
      boxShadow: "shadowFocus",
      color: "colorTextLinkStronger"
    },
    _selected: {
      backgroundColor: "colorBackgroundPrimaryStronger",
      color: "colorTextWeakest"
    }
  }), isHoverable && __assign({}, hoverStyles)), props.children);
});
FormPillStyles.displayName = "StyledFormPill";
var FormPill = React2.forwardRef((_a, ref) => {
  var {element = "FORM_PILL", onDismiss, onSelect, next} = _a, props = __rest(_a, ["element", "onDismiss", "onSelect", "next"]);
  const handleKeydown = React2.useCallback((event) => {
    if ((event.key === "Backspace" || event.key === "Delete") && typeof onDismiss === "function") {
      onDismiss();
      if (typeof next === "function") {
        next();
      }
    }
  }, [onDismiss, next]);
  const isHoverable = onSelect != null;
  return /* @__PURE__ */ React2.createElement(import_reakit_library.CompositeItem, __assign(__assign({
    as: FormPillStyles,
    element
  }, props), {
    ref,
    isHoverable,
    onKeyDown: handleKeydown,
    onClick: onSelect,
    next
  }), props.children, onDismiss != null ? /* @__PURE__ */ React2.createElement(PillCloseIcon, {
    onClick: onDismiss
  }) : null);
});
FormPill.displayName = "FormPill";

// src/FormPillGroup.tsx
var React3 = __toModule(require("react"));
var import_box3 = __toModule(require("@twilio-paste/box"));
var import_reakit_library2 = __toModule(require("@twilio-paste/reakit-library"));
var FormPillGroupStyles = React3.forwardRef((_a, ref) => {
  var {element = "FORM_PILL_GROUP"} = _a, props = __rest(_a, ["element"]);
  return /* @__PURE__ */ React3.createElement(import_box3.Box, __assign(__assign({}, import_box3.safelySpreadBoxProps(props)), {
    element,
    ref,
    role: "listbox",
    margin: "space0",
    padding: "space0",
    display: "flex",
    flexWrap: "wrap",
    rowGap: "space30",
    columnGap: "space30"
  }), props.children);
});
FormPillGroupStyles.displayName = "StyledFormPillGroup";
var FormPillGroup = React3.forwardRef((props, ref) => {
  return /* @__PURE__ */ React3.createElement(import_reakit_library2.Composite, __assign(__assign({
    as: FormPillGroupStyles
  }, props), {
    ref
  }), props.children);
});
FormPillGroup.displayName = "FormPillGroup";

// src/useFormPillState.tsx
var import_reakit_library3 = __toModule(require("@twilio-paste/reakit-library"));
var useFormPillState = (config = {}) => {
  return __assign({}, import_reakit_library3.useCompositeState(__assign(__assign({}, config), {
    orientation: "horizontal",
    loop: true
  })));
};
